JSONWriter.java

```diff
@@ -300,7 +328,7 @@ public JSONWriter value(boolean b) throws JSONException {
      * @throws JSONException If the number is not finite.
      */
     public JSONWriter value(double d) throws JSONException {
-        return this.value(new Double(d));
+        return this.append(JSONObject.numberToString(Double.valueOf(d)));

```

Comment:
Avoid the delegation chain to JSONWriter.value(Object), to JSONObject.writeValue(Object), to JSONObject.numberToString(). Avoid the chain and go there directly. Still retains the testValidity() call.